## Part 2 Comparing Across Classifiers

In [ ]:
from sklearn.datasets import fetch_openml
mnist = fetch_openml('mnist_784', version=1)
mnist.keys()

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


dict_keys(['data', 'target', 'frame', 'categories', 'feature_names', 'target_names', 'DESCR', 'details', 'url'])

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


dict_keys(['data', 'target', 'frame', 'categories', 'feature_names', 'target_names', 'DESCR', 'details', 'url'])

In [ ]:
mnistX = mnist['data']
mnistX.shape

(70000, 784)

(70000, 784)

In [ ]:
mnistY = mnist['target']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(mnistX, mnistY, test_size=0.3, random_state=45)

print(f'Training samples: {X_train.shape[0]:,}')
print(f'Test samples: {X_test.shape[0]:,}')

Training samples: 49,000
Test samples: 21,000
Training samples: 49,000
Test samples: 21,000


In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

For the next section, you will compare the performance of multiple classifiers (tuned via CV) on the test data. You will run these classifiers on the dimension reduced PCA data (NOTE: IN YOUR WRITE-UP, PLEASE GIVE A SHORT TEXT DESCRIPTION OF EACH CLASSIFIER BEFORE YOU USE IT).

- REMEMBER TO TRAIN THE FINAL MODEL ON THE FULL TRAINING SET AFTER TUNING THE PARAMETERS
- I would like you to run the following classifiers using 10-fold CV repeated 5 times to tune the parameters
  - Linear SVM
  - Radial Basis SVM
- I would also like you to train a Random Forest
- I would also like you to train an LDA model
- I would also like you to train a QDA model

For each model, I would like you to compare across both test accuracy and running time. In particular, I would like you to address
- (*)Which model would you choose out of the above (including K-NN)? Why would you make this choice?
- For the models that provide a test error estimate (Random Forest via the error and the CV-tuned models via cross validation), which would you have chosen based on the estimated test error? Explain how this relates to your answer to question (*)?
- For the models that do not immediately provide a test error estimate (LDA and QDA), how could you estimate test error (Hint: CV)?

#### Creating a PCA pipeline for train and test data sets.

We have calculated that 150 components give a 90% explanation for variance on the test data set. We use that here to build the standardized train and test data sets to build different models

In [ ]:
import numpy as np
from sklearn.decomposition import PCA
pca_train = PCA(n_components = 150)
X_train_pca = pca_train.fit_transform(X_train_std)
pca_test = PCA(n_components = 150)
X_test_pca = pca_test.fit_transform(X_test_std)

### Linear and Radial Basis Support Vector Machine

#### Linear SVM

Linear SVM creates linear hyperplanes that attempt to maximize the separation or margin of the classes. All the classes will be maximally separated from each other by a <b>straight line</b>, i.e., each class is defined such that th distance between them is maximized.

The model uses Euclidean distance to determine the distance between the classes.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
pipeline_lsvc = Pipeline([('model',SVC())
                        ])
params_lsvc = [{'model__C':[0.01], 'model__kernel':['linear']}]
lsvc_grid = GridSearchCV(estimator = pipeline_lsvc, param_grid = params_lsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_lsvc = time.time()
lsvc_grid = lsvc_grid.fit(X_train_pca,y_train)
time_elapsed_lsvc = time.time() - start_time_lsvc
accuracy_train_lsvc = lsvc_grid.score(X_train_pca, y_train)
accuracy_test_lsvc = lsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Linear SVC Accuracy Score on reduced test data set: {accuracy_test_lsvc:.2%}")
print(f"Time for model to run on reduced test data set: {time_elapsed_lsvc:.5f} seconds")
print(f"Linear SVC Accuracy Score on reduced train data set: {accuracy_train_lsvc:.2%}")

Linear SVC Accuracy Score on reduced test data set: 4.40%
Time for model to run on reduced test data set: 316.44184 seconds
Linear SVC Accuracy Score on reduced train data set: 94.90%


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
pipeline_lsvc = Pipeline([('model',SVC())
                        ])
params_lsvc = [{'model__C':[0.05], 'model__kernel':['linear']}]
lsvc_grid = GridSearchCV(estimator = pipeline_lsvc, param_grid = params_lsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
start_time_lsvc = time.time()
lsvc_grid = lsvc_grid.fit(X_train_pca,y_train)
time_elapsed_lsvc = time.time() - start_time_lsvc
accuracy_train_lsvc = lsvc_grid.score(X_train_pca, y_train)
accuracy_test_lsvc = lsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Linear SVC Accuracy Score on reduced test data set: {accuracy_test_lsvc:.2%}")
print(f"Time for model to run on reduced test data set: {time_elapsed_lsvc:.5f} seconds")
print(f"Linear SVC Accuracy Score on reduced train data set: {accuracy_train_lsvc:.2%}")

Linear SVC Accuracy Score on reduced test data set: 5.04%
Time for model to run on reduced test data set: 432.12095 seconds
Linear SVC Accuracy Score on reduced train data set: 95.31%


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
pipeline_lsvc = Pipeline([('model',SVC())
                        ])
params_lsvc = [{'model__C':[0.1], 'model__kernel':['linear']}]
lsvc_grid = GridSearchCV(estimator = pipeline_lsvc, param_grid = params_lsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
start_time_lsvc = time.time()
lsvc_grid = lsvc_grid.fit(X_train_pca,y_train)
time_elapsed_lsvc = time.time() - start_time_lsvc
accuracy_train_lsvc = lsvc_grid.score(X_train_pca, y_train)
accuracy_test_lsvc = lsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Linear SVC Accuracy Score on reduced test data set: {accuracy_test_lsvc:.2%}")
print(f"Time for model to run on reduced test data set: {time_elapsed_lsvc:.5f} seconds")
print(f"Linear SVC Accuracy Score on reduced train data set: {accuracy_train_lsvc:.2%}")

Linear SVC Accuracy Score on reduced test data set: 5.39%
Time for model to run on reduced test data set: 436.83987 seconds
Linear SVC Accuracy Score on reduced train data set: 95.48%


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
pipeline_lsvc = Pipeline([('model',SVC())
                        ])
params_lsvc = [{'model__C':[0.15], 'model__kernel':['linear']}]
lsvc_grid = GridSearchCV(estimator = pipeline_lsvc, param_grid = params_lsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
start_time_lsvc = time.time()
lsvc_grid = lsvc_grid.fit(X_train_pca,y_train)
time_elapsed_lsvc = time.time() - start_time_lsvc
accuracy_train_lsvc = lsvc_grid.score(X_train_pca, y_train)
accuracy_test_lsvc = lsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Linear SVC Accuracy Score on reduced test data set: {accuracy_test_lsvc:.2%}")
print(f"Time for model to run on reduced test data set: {time_elapsed_lsvc:.5f} seconds")
print(f"Linear SVC Accuracy Score on reduced train data set: {accuracy_train_lsvc:.2%}")

Linear SVC Accuracy Score on reduced test data set: 5.56%
Time for model to run on reduced test data set: 486.28641 seconds
Linear SVC Accuracy Score on reduced train data set: 95.58%


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
pipeline_lsvc = Pipeline([('model',SVC())
                        ])
params_lsvc = [{'model__C':[0.2], 'model__kernel':['linear']}]
lsvc_grid = GridSearchCV(estimator = pipeline_lsvc, param_grid = params_lsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
start_time_lsvc = time.time()
lsvc_grid = lsvc_grid.fit(X_train_pca,y_train)
time_elapsed_lsvc = time.time() - start_time_lsvc
accuracy_train_lsvc = lsvc_grid.score(X_train_pca, y_train)
accuracy_test_lsvc = lsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Linear SVC Accuracy Score on reduced test data set: {accuracy_test_lsvc:.2%}")
print(f"Time for model to run on reduced test data set: {time_elapsed_lsvc:.5f} seconds")
print(f"Linear SVC Accuracy Score on reduced train data set: {accuracy_train_lsvc:.2%}")

Linear SVC Accuracy Score on reduced test data set: 5.68%
Time for model to run on reduced test data set: 550.90286 seconds
Linear SVC Accuracy Score on reduced train data set: 95.61%


In [ ]:
best_pca_lsvc = 0.15

In [ ]:
lsvc_best = SVC(C = best_pca_lsvc, kernel = 'linear')
start_time_lsvc1 = time.time()
lsvc_best.fit(X_test_pca,y_test)
time_elapsed_lsvc1 = time.time() - start_time_lsvc1
accuracy_test_lsvc1 = lsvc_best.score(X_test_pca, y_test)

In [ ]:
print(f"Time for best model to run on reduced test data set: {time_elapsed_lsvc1:.5f} seconds")
print(f'Linear SVM best Accuracy Score on reduced test data set: {accuracy_test_lsvc1:.2%}')
#print(f'Linear SVM best Accuracy Score on reduced train data set: {accuracy_train_lsvc1:.2%}')

Time for best model to run on reduced test data set: 6.06220 seconds
Linear SVM best Accuracy Score on reduced test data set: 96.35%


> The best model has a c-value of 0.15. It has the best balance of accuracy and runtime.

#### Radial Basis SVM

Radial Basis SVM deals with features with are inherently non-linearly related. i.e., instead of being separated by a straight line, each class will be maximally separated by a curve solved by optimizing the constraints involved.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

In [ ]:
pipeline_rsvc = Pipeline([('model',SVC())
                        ])
params_rsvc = [{'model__C':[0.01], 'model__kernel':['rbf']}]
rsvc_grid = GridSearchCV(estimator = pipeline_rsvc, param_grid = params_rsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_rsvc = time.time()
rsvc_grid = rsvc_grid.fit(X_train_pca,y_train)
time_elapsed_rsvc = time.time() - start_time_rsvc
accuracy_train_rsvc = rsvc_grid.score(X_train_pca, y_train)
accuracy_test_rsvc = rsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Radial Basis SVC Accuracy Score on reduced test data set: {accuracy_test_rsvc:.2%}")
print(f"Time for model to run on reduced traint data set: {time_elapsed_rsvc:.5f} seconds")
print(f"Radial Basis SVC Accuracy Score on reduced train data set: {accuracy_train_rsvc:.2%}")

Radial Basis SVC Accuracy Score on reduced test data set: 8.80%
Time for model to run on reduced traint data set: 3419.57601 seconds
Radial Basis SVC Accuracy Score on reduced train data set: 87.21%


In [ ]:
pipeline_rsvc = Pipeline([('model',SVC())
                        ])
params_rsvc = [{'model__C':[0.05], 'model__kernel':['rbf']}]
rsvc_grid = GridSearchCV(estimator = pipeline_rsvc, param_grid = params_rsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_rsvc = time.time()
rsvc_grid = rsvc_grid.fit(X_train_pca,y_train)
time_elapsed_rsvc = time.time() - start_time_rsvc
accuracy_train_rsvc = rsvc_grid.score(X_train_pca, y_train)
accuracy_test_rsvc = rsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Radial Basis SVC Accuracy Score on reduced test data set: {accuracy_test_rsvc:.2%}")
print(f"Time for model to run on reduced traint data set: {time_elapsed_rsvc:.5f} seconds")
print(f"Radial Basis SVC Accuracy Score on reduced train data set: {accuracy_train_rsvc:.2%}")

Radial Basis SVC Accuracy Score on reduced test data set: 7.29%
Time for model to run on reduced traint data set: 1745.66063 seconds
Radial Basis SVC Accuracy Score on reduced train data set: 92.88%


In [ ]:
pipeline_rsvc = Pipeline([('model',SVC())
                        ])
params_rsvc = [{'model__C':[0.1], 'model__kernel':['rbf']}]
rsvc_grid = GridSearchCV(estimator = pipeline_rsvc, param_grid = params_rsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_rsvc = time.time()
rsvc_grid = rsvc_grid.fit(X_train_pca,y_train)
time_elapsed_rsvc = time.time() - start_time_rsvc
accuracy_train_rsvc = rsvc_grid.score(X_train_pca, y_train)
accuracy_test_rsvc = rsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Radial Basis SVC Accuracy Score on reduced test data set: {accuracy_test_rsvc:.2%}")
print(f"Time for model to run on reduced traint data set: {time_elapsed_rsvc:.5f} seconds")
print(f"Radial Basis SVC Accuracy Score on reduced train data set: {accuracy_train_rsvc:.2%}")

Radial Basis SVC Accuracy Score on reduced test data set: 7.07%
Time for model to run on reduced traint data set: 1295.54594 seconds
Radial Basis SVC Accuracy Score on reduced train data set: 94.47%


In [ ]:
pipeline_rsvc = Pipeline([('model',SVC())
                        ])
params_rsvc = [{'model__C':[0.15], 'model__kernel':['rbf']}]
rsvc_grid = GridSearchCV(estimator = pipeline_rsvc, param_grid = params_rsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_rsvc = time.time()
rsvc_grid = rsvc_grid.fit(X_train_pca,y_train)
time_elapsed_rsvc = time.time() - start_time_rsvc
accuracy_train_rsvc = rsvc_grid.score(X_train_pca, y_train)
accuracy_test_rsvc = rsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Radial Basis SVC Accuracy Score on reduced test data set: {accuracy_test_rsvc:.2%}")
print(f"Time for model to run on reduced traint data set: {time_elapsed_rsvc:.5f} seconds")
print(f"Radial Basis SVC Accuracy Score on reduced train data set: {accuracy_train_rsvc:.2%}")

Radial Basis SVC Accuracy Score on reduced test data set: 7.04%
Time for model to run on reduced traint data set: 1074.42384 seconds
Radial Basis SVC Accuracy Score on reduced train data set: 95.23%


In [ ]:
pipeline_rsvc = Pipeline([('model',SVC())
                        ])
params_rsvc = [{'model__C':[0.2], 'model__kernel':['rbf']}]
rsvc_grid = GridSearchCV(estimator = pipeline_rsvc, param_grid = params_rsvc, scoring = 'accuracy', refit = True, cv = 10)

In [ ]:
import time
start_time_rsvc = time.time()
rsvc_grid = rsvc_grid.fit(X_train_pca,y_train)
time_elapsed_rsvc = time.time() - start_time_rsvc
accuracy_train_rsvc = rsvc_grid.score(X_train_pca, y_train)
accuracy_test_rsvc = rsvc_grid.score(X_test_pca, y_test)

In [ ]:
print(f"Radial Basis SVC Accuracy Score on reduced test data set: {accuracy_test_rsvc:.2%}")
print(f"Time for model to run on reduced traint data set: {time_elapsed_rsvc:.5f} seconds")
print(f"Radial Basis SVC Accuracy Score on reduced train data set: {accuracy_train_rsvc:.2%}")

Radial Basis SVC Accuracy Score on reduced test data set: 7.05%
Time for model to run on reduced traint data set: 1038.79500 seconds
Radial Basis SVC Accuracy Score on reduced train data set: 95.70%


In [ ]:
best_pca_rsvc = 0.2

In [ ]:
rsvc_best = SVC(C = best_pca_rsvc, kernel = 'rbf')
start_time_rsvc1 = time.time()
rsvc_best.fit(X_test_pca,y_test)
time_elapsed_rsvc1 = time.time() - start_time_rsvc1
accuracy_test_rsvc1 = rsvc_best.score(X_test_pca, y_test)
accuracy_train_rsvc1 = rsvc_best.score(X_train_pca, y_train)

In [ ]:
print(f"Time for best model to run on reduced test data set: {time_elapsed_rsvc1:.5f} seconds")
print(f'Radial Basis SVM best Accuracy Score on reduced test data set: {accuracy_test_rsvc1:.2%}')
print(f'Radial Basis SVM best Accuracy Score on reduced train data set: {accuracy_train_rsvc1:.2%}')

Time for best model to run on reduced test data set: 13.86952 seconds
Radial Basis SVM best Accuracy Score on reduced test data set: 93.66%
Radial Basis SVM best Accuracy Score on reduced train data set: 5.89%


> The best model has a c-value of 0.2 with the best optimization of accuracy and runtime

### Random Forest Classifiers

Random Forests are complicated and evolved form of decision trees. They involve multiple decision trees at each node creating a strong learner and handles data imbalances more efficiently.

It votes the class for each data point based on the decision made by the trees as defined by the model builder.

This model reduces the effect of noise on the evaluation of the model.

Tuning Parameters -
- Number of trees
- Number of samples in each class
- Samples per leaf
- Samples per split
- Depth of the trees

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators = 100, random_state = 42)

In [ ]:
import time
start_time_rf = time.time()
rf_model = rf_model.fit(X_train_pca,y_train)
time_elapsed_rf = time.time() - start_time_rf
accuracy_train_rf = rf_model.score(X_train_pca, y_train)
accuracy_test_rf = rf_model.score(X_test_pca, y_test)

In [ ]:
print(f"Random Forest Accuracy Score on reduced train data set: {accuracy_train_rf:.2%}")
print(f"Random Forest Accuracy Score on reduced test data set: {accuracy_test_rf:.2%}")
print(f"Time for model to run on reduced train data set: {time_elapsed_rf:.5f} seconds")

Random Forest Accuracy Score on reduced train data set: 100.00%
Random Forest Accuracy Score on reduced test data set: 6.47%
Time for model to run on reduced train data set: 87.29230 seconds


In [ ]:
from sklearn.metrics import accuracy_score
rf_y_pred = rf_model.predict(X_test_pca)
rf_test_accuracy = accuracy_score(y_test, rf_y_pred)
rf_test_error = 1 - accuracy_test_rf
print(f'The test error estimate for Random Forest Classifier model is: {rf_test_error:.2%}')

The test error estimate for Random Forest Classifier model is: 93.53%


### Linear Discriminant Analysis (LDA)

LDA is a supervised learning algorithm that finds axes or hyperplanes that distinguish the most between classes.

- The distinction from PCA is that the for LDA classes are as far as possible.
- LDA assumes the data is normally distributed.
- Similar to logistic regression as the samples are independenlty treated unlike in PCA
> LDA involves calculation of eigenvalues and ranking their respective eigenvectors. k eigenvectors are picked to project.

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

lda = LDA(n_components = 2)
X_train_lda = lda.fit_transform(X_train_std, y_train)

In [ ]:
from sklearn.model_selection import cross_val_score
train_lda_score = cross_val_score(lda,X_train_std, y_train, cv = 5, scoring = 'accuracy')
test_lda_score = cross_val_score(lda,X_test_std, y_test, cv = 5, scoring = 'accuracy')

In [ ]:
print(f'Average Accuracy Score of train data using LDA: {train_lda_score.mean():.2%}')
print(f'Average Accuracy Score of test data using LDA: {test_lda_score.mean():.2%}')

Average Accuracy Score of train data using LDA: 81.49%
Average Accuracy Score of test data using LDA: 9.86%


### Quadratic Discriminant Analysis (QDA)

QDA invloves quadratic decision boundaries or hyperplanes unline linear onea in LDA. It allows flexibility in shape of the boundary shared by different classes and therefore could be computationally more exhausting.

In [ ]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
qda = QuadraticDiscriminantAnalysis()
X_train_qda = qda.fit(X_train_std, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


In [ ]:
train_qda_score = cross_val_score(qda,X_train_std, y_train, cv = 5, scoring = 'accuracy')
test_qda_score = cross_val_score(qda,X_test_std, y_test,cv = 5, scoring = 'accuracy')

/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/discriminant_analysis.py:926: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
/usr/local/lib/python3.10/dist-packages/sklearn/disc

In [ ]:
print(f'Average Accuracy Score of train data using QDA: {train_qda_score.mean():.2%}')
print(f'Average Accuracy Score of test data using QDA: {test_qda_score.mean():.2%}')

Average Accuracy Score of train data using QDA: 51.83%
Average Accuracy Score of test data using QDA: 57.71%


### Summary of Models
|    Model    |  Train Accuracy  |  Test Accuracy  |  Time for execution  |
|---------|------------------|----------|---------------------|
|     KNN     |  97.11%   |  94.25%   |   0.10764s  |
| Linear SVC  |  95.58%   |   5.56%   |   6.0266s   |
| Radial SVC  |  95.70%   |   7.05%   |   13.8695s  |
|Random Forest| 100.00%   |   6.47%   |   -         |
|LDA          |  81.49%   |   9.86%   |  -         |
|QDA          |  51.83%   |   57.71%  |   -         |


The above test accuracies are obtained after fitting the models with the train data set.
- The KNN model has the highest test data accuracy. This is because 250 components were used to fit the train data set, whereas for other models 150 were used.
- Random Forest Classifier has the highest train data accuracy.
- All the models except KNN and QDA overfits the train data as the test accuracy is much lesser than the train accuracy.
- QDA has a relatable train and test accuracy, however, it is not adequate to consider for further modeling or prediction. The reason being that the components considered have a high level of collinearity.
> <b>The K-Nearest Neighbor model works the best among all the classification models used in this project.</b>